File: verify_links.js
const fs = require('fs');
const path = require('path');

const ROOT_DIR = __dirname;
console.log(`Auditing repository at: ${ROOT_DIR}\n`);

const htmlFiles = fs.readdirSync(ROOT_DIR).filter(f => f.endsWith('.html'));

const errors = [];
const warnings = [];
const hreflangMappings = {};  // {file: {lang: url}}
const canonicalMappings = {}; // {file: url}
const GA4_ID = 'G-FXXM8ZF82P';

// Keep track of which pages are IT and which are DE
const itPages = [
    'index.html', 'chi-siamo.html', 'portfolio.html', 
    'ristrutturazione-bagno-merano.html', 'impermeabilizzazioni-bolzano.html', 
    'grazie.html', 'impressum.html', 'privacy-policy.html', 'cookie-policy.html'
];
const dePages = [
    'index-de.html', 'ueber-uns.html', 'projektbeispiele.html', 
    'badsanierung-meran.html', 'abdichtungen-bozen.html', 
    'vielen-dank.html', 'impressum-de.html', 'privacy-policy-de.html', 'cookie-policy-de.html'
];

for (const file of htmlFiles) {
    const filePath = path.join(ROOT_DIR, file);
    const content = fs.readFileSync(filePath, 'utf-8');

    console.log(`=== Checking ${file} ===`);
    
    // 1. Check <html lang="...">
    const langMatch = content.match(/<html\s+lang=["']([^"']+)["']/i);
    let lang = '';
    if (!langMatch) {
        errors.push(`[${file}] Missing <html lang='...'> attribute.`);
    } else {
        lang = langMatch[1];
        console.log(`  - Language: ${lang}`);
        if (lang !== 'it' && lang !== 'de') {
            warnings.push(`[${file}] Language attribute '${lang}' is not 'it' or 'de'.`);
        }
    }

    // 2. Check GA4 tracking script
    const gaScriptPresent = content.includes(`https://www.googletagmanager.com/gtag/js?id=${GA4_ID}`);
    const gaConfigPresent = content.includes(`gtag('config', '${GA4_ID}')`) || content.includes(`gtag("config", "${GA4_ID}")`);
    
    if (!gaScriptPresent) {
        errors.push(`[${file}] GA4 tracking script with ID ${GA4_ID} not found in HTML head.`);
    }
    if (!gaConfigPresent) {
        errors.push(`[${file}] GA4 gtag('config', '${GA4_ID}') initialization not found.`);
    }

    // 3. Check Canonical and Alternate Links
    if (file !== '404.html') {
        const canonicalMatch = content.match(/<link\s+rel=["']canonical["']\s+href=["']([^"']+)["']/i);
        if (!canonicalMatch) {
            errors.push(`[${file}] Missing <link rel='canonical'> tag.`);
        } else {
            canonicalMappings[file] = canonicalMatch[1];
        }

        const alternates = [];
        const altRegex1 = /<link\s+rel=["']alternate["']\s+hreflang=["']([^"']+)["']\s+href=["']([^"']+)["']/gi;
        const altRegex2 = /<link\s+hreflang=["']([^"']+)["']\s+rel=["']alternate["']\s+href=["']([^"']+)["']/gi;
        
        let match;
        while ((match = altRegex1.exec(content)) !== null) {
            alternates.push({ lang: match[1], href: match[2] });
        }
        while ((match = altRegex2.exec(content)) !== null) {
            alternates.push({ lang: match[1], href: match[2] });
        }
        
        hreflangMappings[file] = {};
        for (const alt of alternates) {
            hreflangMappings[file][alt.lang] = alt.href;
        }
            
        console.log(`  - Alternates:`, hreflangMappings[file]);
        if (!hreflangMappings[file]['it']) {
            errors.push(`[${file}] Missing hreflang='it' alternate link.`);
        }
        if (!hreflangMappings[file]['de']) {
            errors.push(`[${file}] Missing hreflang='de' alternate link.`);
        }
    }

    // 4. Check Internal Links
    const anchorRegex = /<a\s+[^>]*href=["']([^"']+)["']/gi;
    let anchorMatch;
    while ((anchorMatch = anchorRegex.exec(content)) !== null) {
        const link = anchorMatch[1];
        
        // Ignore external links, mailto, tel, hashes
        if (link.startsWith('http://') || link.startsWith('https://') || link.startsWith('mailto:') || link.startsWith('tel:') || link.startsWith('#')) {
            if (link.includes('idalbau.github.io')) {
                const parts = link.split('idalbau.github.io/');
                const pathPart = parts[parts.length - 1].split('#')[0];
                if (pathPart && !htmlFiles.includes(pathPart) && pathPart !== '') {
                    errors.push(`[${file}] Link to domain page '${link}' points to non-existent local file '${pathPart}'.`);
                }
            }
            continue;
        }
        
        const localFile = link.split('?')[0].split('#')[0];
        if (!localFile) continue;
            
        const localPath = path.join(ROOT_DIR, localFile);
        if (!fs.existsSync(localPath)) {
            errors.push(`[${file}] Link '${link}' points to non-existent file '${localFile}'.`);
        }
    }

    // 5. Check Forms and Redirects
    const formRegex = /<form\s+[^>]*action=["']([^"']+)["']/gi;
    let formMatch;
    while ((formMatch = formRegex.exec(content)) !== null) {
        const formAction = formMatch[1];
        if (formAction.includes('web3forms.com')) {
            const redirectRegex = /<input\s+type=["']hidden["']\s+name=["']redirect["']\s+value=["']([^"']+)["']/i;
            const redirectMatch = content.match(redirectRegex);
            if (!redirectMatch) {
                errors.push(`[${file}] Form submits to Web3Forms but has no redirect URL.`);
            } else {
                const redirectUrl = redirectMatch[1];
                const expectedRedirect = lang === 'it' ? 'grazie.html' : 'vielen-dank.html';
                if (!redirectUrl.includes(expectedRedirect)) {
                    errors.push(`[${file}] Form redirect is '${redirectUrl}', expected to contain '${expectedRedirect}' for language '${lang}'.`);
                }
            }
            
            // Check for Privacy Policy link matching form language
            const privacyRegex = /<a\s+[^>]*href=["']([^"']+)["'][^>]*>Privacy Policy<\/a>/i;
            const privacyMatch = content.match(privacyRegex);
            if (privacyMatch) {
                const privacyLink = privacyMatch[1];
                const expectedPrivacy = lang === 'it' ? 'privacy-policy.html' : 'privacy-policy-de.html';
                if (privacyLink !== expectedPrivacy) {
                    errors.push(`[${file}] GDPR form links to privacy policy '${privacyLink}', but expected '${expectedPrivacy}' for language '${lang}'.`);
                }
            }
        }
    }

    // 6. Check WhatsApp Float Button Widget
    const waRegex = /<a\s+[^>]*href=["']https:\/\/wa\.me\/[^"']+["']([^>]*)>/gi;
    let waMatch;
    while ((waMatch = waRegex.exec(content)) !== null) {
        const tagAttributes = waMatch[1];
        if (!tagAttributes.includes('target="_blank"')) {
            errors.push(`[${file}] WhatsApp float link missing target='_blank'.`);
        }
        if (!tagAttributes.includes('rel="noopener noreferrer"') && !tagAttributes.includes('rel="noreferrer noopener"')) {
            errors.push(`[${file}] WhatsApp float link missing rel='noopener noreferrer'.`);
        }
        if (!tagAttributes.includes('gtag') && !tagAttributes.includes('onclick')) {
            errors.push(`[${file}] WhatsApp float link missing onClick GA4 tracking event.`);
        }
        
        // Find text content inside the tag
        const escapedWaHref = waMatch[0].replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
        const textRegex = new RegExp(`<a\\s+[^>]*href=["']https:\\/\\/wa\\.me\\/[^"']+["'][^>]*>([\\s\\S]*?)<\\/a>`, 'i');
        const waFullMatch = content.match(textRegex);
        if (waFullMatch) {
            const innerText = waFullMatch[1].toLowerCase();
            if (lang === 'it' && !innerText.includes('scrivici')) {
                warnings.push(`[${file}] WhatsApp text for IT page doesn't contain 'Scrivici' (got: '${innerText.trim()}').`);
            } else if (lang === 'de' && !innerText.includes('schreiben') && !innerText.includes('kontakt')) {
                warnings.push(`[${file}] WhatsApp text for DE page doesn't contain 'Schreiben Sie uns' (got: '${innerText.trim()}').`);
            }
        }
    }
}

// 7. Check Bidirectional Hreflang Alternates
console.log('\n=== Verifying Bidirectional Alternates ===');
for (const [file, mapping] of Object.entries(hreflangMappings)) {
    // 404 pages or policy pages might not be mapped in alternates but standard ones should be
    // Actually, it's good to check standard pages
    for (const [altLang, altUrl] of Object.entries(mapping)) {
        const parts = altUrl.split('idalbau.github.io/');
        let targetFilename = parts[parts.length - 1].split('#')[0];
        if (!targetFilename || targetFilename === '') {
            targetFilename = 'index.html';
        }
            
        if (!htmlFiles.includes(targetFilename)) {
            errors.push(`[${file}] Alternate URL '${altUrl}' maps to non-existent local file '${targetFilename}'.`);
            continue;
        }
            
        const targetMapping = hreflangMappings[targetFilename] || {};
        const sourceLang = itPages.includes(file) ? 'it' : 'de';
        
        const backLink = targetMapping[sourceLang] || '';
        const backParts = backLink.split('idalbau.github.io/');
        let backFilename = backParts[backParts.length - 1].split('#')[0];
        if (!backFilename || backFilename === '') {
            backFilename = 'index.html';
        }
            
        if (backFilename !== file) {
            errors.push(`[${file}] Alternate mapping to '${targetFilename}' is not bidirectional. '${targetFilename}' alternates map back to '${backFilename}' instead of '${file}'.`);
        }
    }
}

// 8. Check Sitemap.xml validity
const sitemapPath = path.join(ROOT_DIR, 'sitemap.xml');
if (fs.existsSync(sitemapPath)) {
    console.log('\n=== Auditing sitemap.xml ===');
    try {
        const sitemapContent = fs.readFileSync(sitemapPath, 'utf-8');
            
        const expectedPagesInSitemap = [
            'index.html', 'chi-siamo.html', 'portfolio.html', 'ristrutturazione-bagno-merano.html', 'impermeabilizzazioni-bolzano.html',
            'index-de.html', 'ueber-uns.html', 'projektbeispiele.html', 'badsanierung-meran.html', 'abdichtungen-bozen.html'
        ];
        
        for (const p of expectedPagesInSitemap) {
            const pClean = p === 'index.html' ? '' : p;
            if (pClean === '') {
                if (!sitemapContent.includes('https://idalbau.github.io/')) {
                    errors.push(`[sitemap.xml] Missing reference to root URL https://idalbau.github.io/.`);
                }
            } else {
                if (!sitemapContent.includes(`https://idalbau.github.io/${p}`)) {
                    errors.push(`[sitemap.xml] Missing reference to page https://idalbau.github.io/${p}.`);
                }
            }
        }
                    
        // Check alternate links in sitemap.xml
        const alternateTagsCount = (sitemapContent.match(/xhtml:link/g) || []).length;
        console.log(`  - Found ${alternateTagsCount} alternate page links in sitemap.xml`);
        if (alternateTagsCount === 0) {
            errors.push('[sitemap.xml] Missing xhtml:link alternate tags.');
        }
            
    } catch (e) {
        errors.push(`Failed to parse sitemap.xml: ${e.message}`);
    }
} else {
    errors.push('Missing sitemap.xml in root directory!');
}

console.log('\n=== AUDIT RESULTS ===');
console.log(`Total Errors found: ${errors.length}`);
console.log(`Total Warnings found: ${warnings.length}`);

if (errors.length > 0) {
    console.log('\nERRORS:');
    errors.forEach(err => console.log(`❌ ${err}`));
}

if (warnings.length > 0) {
    console.log('\nWARNINGS:');
    warnings.forEach(warn => console.log(`⚠️ ${warn}`));
}

if (errors.length === 0 && warnings.length === 0) {
    console.log('\n✅ All checks passed! No errors or warnings found. Excellent work!');
} else if (errors.length === 0) {
    console.log('\n✅ No blocker errors found! Just minor warnings.');
} else {
    process.exit(1);
}

File: verify_links.py
import os
import re
import xml.etree.ElementTree as ET

# Root directory of the repository
ROOT_DIR = r"c:\Users\daniel\Documents\GitHub\idalbau.github.io"
# Fallback to current directory if not found (in case of path nuances)
if not os.path.exists(ROOT_DIR):
    ROOT_DIR = os.path.dirname(os.path.dirname(os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))))
    if not os.path.exists(os.path.join(ROOT_DIR, "index.html")):
        ROOT_DIR = os.getcwd()

print(f"Auditing repository at: {ROOT_DIR}\n")

# Find all HTML files
html_files = [f for f in os.listdir(ROOT_DIR) if f.endswith(".html")]

errors = []
warnings = []
hreflang_mappings = {}  # {file: {lang: url}}
canonical_mappings = {} # {file: url}

# GA4 measurement ID to verify
GA4_ID = "G-FXXM8ZF82P"

for file in html_files:
    file_path = os.path.join(ROOT_DIR, file)
    with open(file_path, "r", encoding="utf-8") as f:
        content = f.read()

    print(f"=== Checking {file} ===")
    
    # 1. Check <html lang="...">
    lang_match = re.search(r'<html\s+lang=["\']([^"\']+)["\']', content, re.IGNORECASE)
    if not lang_match:
        errors.append(f"[{file}] Missing <html lang='...'> attribute.")
    else:
        lang = lang_match.group(1)
        print(f"  - Language: {lang}")
        if lang not in ["it", "de"]:
            warnings.append(f"[{file}] Language attribute '{lang}' is not 'it' or 'de'.")

    # 2. Check GA4 tracking script
    ga_script_present = f"https://www.googletagmanager.com/gtag/js?id={GA4_ID}" in content
    ga_config_present = f"gtag('config', '{GA4_ID}')" in content or f'gtag("config", "{GA4_ID}")' in content
    
    # Skip thank you page or policy if they don't have standard GA4 but check anyways
    if not ga_script_present:
        errors.append(f"[{file}] GA4 tracking script with ID {GA4_ID} not found in HTML head.")
    if not ga_config_present:
        errors.append(f"[{file}] GA4 gtag('config', '{GA4_ID}') initialization not found.")

    # 3. Check Canonical and Alternate Links
    canonical_match = re.search(r'<link\s+rel=["\']canonical["\']\s+href=["\']([^"\']+)["\']', content, re.IGNORECASE)
    if not canonical_match:
        errors.append(f"[{file}] Missing <link rel='canonical'> tag.")
    else:
        canonical_mappings[file] = canonical_match.group(1)

    alternates = re.findall(r'<link\s+rel=["\']alternate["\']\s+hreflang=["\']([^"\']+)["\']\s+href=["\']([^"\']+)["\']', content, re.IGNORECASE)
    # Check alternate format as well where attributes are swapped
    alternates_swapped = re.findall(r'<link\s+hreflang=["\']([^"\']+)["\']\s+rel=["\']alternate["\']\s+href=["\']([^"\']+)["\']', content, re.IGNORECASE)
    all_alternates = alternates + alternates_swapped
    
    hreflang_mappings[file] = {}
    for lang, href in all_alternates:
        hreflang_mappings[file][lang] = href
        
    print(f"  - Alternates: {hreflang_mappings[file]}")
    if "it" not in hreflang_mappings[file]:
        errors.append(f"[{file}] Missing hreflang='it' alternate link.")
    if "de" not in hreflang_mappings[file]:
        errors.append(f"[{file}] Missing hreflang='de' alternate link.")

    # 4. Check Internal Links
    links = re.findall(r'<a\s+[^>]*href=["\']([^"\']+)["\']', content, re.IGNORECASE)
    for link in links:
        # Ignore external links, mailto, tel, hashes
        if link.startswith(("http://", "https://", "mailto:", "tel:", "#")):
            # If it is self domain, check it
            if "idalbau.github.io" in link:
                path_part = link.split("idalbau.github.io/")[-1].split("#")[0]
                if path_part and path_part not in html_files and path_part != "":
                    errors.append(f"[{file}] Link to domain page '{link}' points to non-existent local file '{path_part}'.")
            continue
        
        # Strip query params or hash
        local_file = link.split("?")[0].split("#")[0]
        if not local_file:
            continue
            
        local_path = os.path.join(ROOT_DIR, local_file)
        if not os.path.exists(local_path):
            errors.append(f"[{file}] Link '{link}' points to non-existent file '{local_file}'.")

    # 5. Check Forms and Redirects
    form_matches = re.findall(r'<form\s+[^>]*action=["\']([^"\']+)["\']', content, re.IGNORECASE)
    for form_action in form_matches:
        # If it's a contact form submitting to web3forms, verify redirect
        if "web3forms.com" in form_action:
            redirect_match = re.search(r'<input\s+type=["\']hidden["\']\s+name=["\']redirect["\']\s+value=["\']([^"\']+)["\']', content, re.IGNORECASE)
            if not redirect_match:
                errors.append(f"[{file}] Form submits to Web3Forms but has no redirect URL.")
            else:
                redirect_url = redirect_match.group(1)
                expected_redirect = "grazie.html" if lang == "it" else "vielen-dank.html"
                if expected_redirect not in redirect_url:
                    errors.append(f"[{file}] Form redirect is '{redirect_url}', expected to contain '{expected_redirect}' for language '{lang}'.")
            
            # Check for Privacy Policy link matching form language
            privacy_link_match = re.search(r'<a\s+[^>]*href=["\']([^"\']+)["\'][^>]*>Privacy Policy</a>', content, re.IGNORECASE)
            if privacy_link_match:
                privacy_link = privacy_link_match.group(1)
                expected_privacy = "privacy-policy.html" if lang == "it" else "privacy-policy-de.html"
                if privacy_link != expected_privacy:
                    errors.append(f"[{file}] GDPR form links to privacy policy '{privacy_link}', but expected '{expected_privacy}' for language '{lang}'.")

    # 6. Check WhatsApp Float Button Widget
    # Search for wa.me link
    wa_match = re.search(r'<a\s+[^>]*href=["\']https://wa.me/[^"\']+["\'][^>]*>', content, re.IGNORECASE)
    if wa_match:
        tag_content = wa_match.group(0)
        if 'target="_blank"' not in tag_content:
            errors.append(f"[{file}] WhatsApp float link missing target='_blank'.")
        if 'rel="noopener noreferrer"' not in tag_content and 'rel="noreferrer noopener"' not in tag_content:
            errors.append(f"[{file}] WhatsApp float link missing rel='noopener noreferrer'.")
        if 'gtag' not in tag_content and 'onclick' not in tag_content:
            errors.append(f"[{file}] WhatsApp float link missing onClick GA4 tracking event.")
            
        # Check text content inside whatsapp tag
        wa_full_tag = re.search(r'<a\s+[^>]*href=["\']https://wa.me/[^"\']+["\'][^>]*>(.*?)</a>', content, re.DOTALL | re.IGNORECASE)
        if wa_full_tag:
            inner_text = wa_full_tag.group(1).lower()
            if lang == "it" and "scrivici" not in inner_text:
                warnings.append(f"[{file}] WhatsApp text for IT page doesn't contain 'Scrivici' (got: '{inner_text.strip()}').")
            elif lang == "de" and "schreiben" not in inner_text and "kontakt" not in inner_text:
                warnings.append(f"[{file}] WhatsApp text for DE page doesn't contain 'Schreiben Sie uns' (got: '{inner_text.strip()}').")

# 7. Check Bidirectional Hreflang Alternates
print("\n=== Verifying Bidirectional Alternates ===")
for file, mapping in hreflang_mappings.items():
    for lang, url in mapping.items():
        # Get target filename from URL
        target_filename = url.split("idalbau.github.io/")[-1].split("#")[0]
        if not target_filename or target_filename == "":
            target_filename = "index.html"
            
        if target_filename not in html_files:
            errors.append(f"[{file}] Alternate URL '{url}' maps to non-existent local file '{target_filename}'.")
            continue
            
        # Check if target page maps back to this file
        target_mapping = hreflang_mappings.get(target_filename, {})
        source_lang = "it" if file in ["index.html", "chi-siamo.html", "portfolio.html", "ristrutturazione-bagno-merano.html", "impermeabilizzazioni-bolzano.html", "grazie.html", "impressum.html", "privacy-policy.html", "cookie-policy.html"] else "de"
        
        # Special check for root mapping (which defaults to IT)
        if file == "index.html":
            source_lang = "it"
            
        back_link = target_mapping.get(source_lang, "")
        back_filename = back_link.split("idalbau.github.io/")[-1].split("#")[0]
        if not back_filename or back_filename == "":
            back_filename = "index.html"
            
        if back_filename != file:
            errors.append(f"[{file}] Alternate mapping to '{target_filename}' is not bidirectional. '{target_filename}' alternates map back to '{back_filename}' instead of '{file}'.")

# 8. Check Sitemap.xml validity
sitemap_path = os.path.join(ROOT_DIR, "sitemap.xml")
if os.path.exists(sitemap_path):
    print("\n=== Auditing sitemap.xml ===")
    try:
        # Simple parse or regex to verify sitemap elements
        with open(sitemap_path, "r", encoding="utf-8") as sf:
            sitemap_content = sf.read()
            
        # Check if all 10 pages are included
        expected_pages_in_sitemap = [
            "index.html", "chi-siamo.html", "portfolio.html", "ristrutturazione-bagno-merano.html", "impermeabilizzazioni-bolzano.html",
            "index-de.html", "ueber-uns.html", "projektbeispiele.html", "badsanierung-meran.html", "abdichtungen-bozen.html"
        ]
        
        for p in expected_pages_in_sitemap:
            p_clean = "" if p == "index.html" else p
            if p_clean == "":
                if "https://idalbau.github.io/" not in sitemap_content:
                    errors.append(f"[sitemap.xml] Missing reference to root URL https://idalbau.github.io/.")
            else:
                if f"https://idalbau.github.io/{p}" not in sitemap_content:
                    errors.append(f"[sitemap.xml] Missing reference to page https://idalbau.github.io/{p}.")
                    
        # Check alternate links in sitemap.xml
        alternate_tags_count = sitemap_content.count("xhtml:link")
        print(f"  - Found {alternate_tags_count} alternate page links in sitemap.xml")
        if alternate_tags_count == 0:
            errors.append("[sitemap.xml] Missing xhtml:link alternate tags.")
            
    except Exception as e:
        errors.append(f"Failed to parse sitemap.xml: {str(e)}")
else:
    errors.append("Missing sitemap.xml in root directory!")

print("\n=== AUDIT RESULTS ===")
print(f"Total Errors found: {len(errors)}")
print(f"Total Warnings found: {len(warnings)}")

if errors:
    print("\nERRORS:")
    for err in errors:
        print(f"❌ {err}")

if warnings:
    print("\nWARNINGS:")
    for warn in warnings:
        print(f"⚠️ {warn}")

if not errors and not warnings:
    print("\n✅ All checks passed! No errors or warnings found. Excellent work!")
elif not errors:
    print("\n✅ No blocker errors found! Just minor warnings.")
